### PR TITLE
Automates cluster upgrade test

### LINKED ISSUE #560: Automates cluster upgrade test
Automates cluster upgrade test not in CI, just for manual testing
The test must accept source and target version parameters, checkout source branch or tag, deploys operator version, create a cluster and than upgrade it to the target version.
#575 need to be resolved before
#576 need to be resolved before
#579 need to be resolved before

### PR DESCRIPTION
**Pull request by dmvolod**
Fixes #560

### CODE REVIEW

**Comment by ryanemerson** [comment]
@dmvolod Can you add some instructions on how to run this to the Readme

**Comment by dmvolod** [comment]
> @dmvolod Can you add some instructions on how to run this to the Readme

Yes, sure, @ryanemerson 
As it's just make file command, added required parameters to the `make` output for upgrade-test and upgrade-test-local targets

**Comment by rigazilla** [comment]
@dmvolod is this a correct cmd line to run the test?
`make upgrade-test-local FROM_UPGRADE_VERSION=2.0.x TO_UPGRADE_VERSION=master`

**Comment by dmvolod** [comment]
> @dmvolod is this a correct cmd line to run the test?
> `make upgrade-test-local FROM_UPGRADE_VERSION=2.0.x TO_UPGRADE_VERSION=master`

Yes, @rigazilla but will not work as master don't have specific changes yet. You can only upgrade to the branch where this PR has been merged

**Comment by dmvolod** [comment]
> Uh yeah.
> This will succeeded then?
> 
> `TO_UPGRADE_VERSION=issue-560`

Yes, sure. I'm using this option. But must be failed due to #579

**Review by rigazilla — COMMENTED** [review]
Test runs fine locally on my PC. 
Only issue: infinispan CR is not deleted at the end
```
PASS
ok      github.com/infinispan/infinispan-operator/test/e2e      92.359s
Already on 'issue-560-fix'
✔ ~/go/src/github.com/infinispan/infinispan-operator [issue-560-fix L|…35418⚑ 66] 
18:13 $ oc get infinispan
NAME                    AGE
test-operator-upgrade   2m38s
```

**Comment by dmvolod** [comment]
@rigazilla it's very strange
got following error in operator log
```
{"level":"error","ts":1601914929.4117632,"logger":"controller_infinispan","msg":"failed to configure new StatefulSet","Request.Namespace":"infinispan-operator","Request.Name":"test-operator-upgrade","error":"quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/dvolodin/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/infinispan/infinispan-operator/pkg/controller/infinispan.(*ReconcileInfinispan).Reconcile\n\t/u01/git/infinispan/infinispan-operator/pkg/controller/infinispan/infinispan_controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/home/dvolodin/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/home/dvolodin/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"}
```
Probably different OCP versions or anything else. Or some old versions are running

**Comment by dmvolod** [comment]
Full upgrade log
[upgrade-2.0.x-to-issue-560.log](https://github.com/infinispan/infinispan-operator/files/5328712/upgrade-2.0.x-to-issue-560.log)

**Comment by dmvolod** [comment]
@rigazilla have you fetch latest version from PR

**Comment by rigazilla** [comment]
> @rigazilla have you fetch latest version from PR

@dmvolod, sorry I wasn't clear. I've added your fix #579 locally on my PC, that's way test doesn't fail. So I think your test is ok.

Just seems it doesn't delete the Infinispan CR on exit.

**Comment by dmvolod** [comment]
> @dmvolod, sorry I wasn't clear. I've added your fix #579 locally on my PC, that's way test doesn't fail. So I think your test is ok.

Yes, correct. I'm speaking about `git fetch upstream pull/583/head:pr-583` to fetch PR from the github
What is you test env? If it's kubernetes the fails doesn't occurs as some k8s versions are not validating OpenAPI Schema
This fail only happens on OCP 4.x

**Comment by dmvolod** [comment]
> Just seems it doesn't delete the Infinispan CR on exit.

Yes, sure need to be added the final clean-up. But strange that you can't reproduce the fails test

**Comment by rigazilla** [comment]
> > Just seems it doesn't delete the Infinispan CR on exit.
> 
> Yes, sure need to be added the final clean-up. But strange that you can't reproduce the fails test

I can reproduce the fail with the clean PR

**Comment by dmvolod** [comment]
@rigazilla added clean-up, but would be nice to have a special flag which will keep all CR's in cluster after upgrade for further problem detection

**Comment by dmvolod** [comment]
@rigazilla have you tested this set of scripts for upgrade or something need to improve?

**Comment by rigazilla** [comment]
I've added a commit to fix cleanup

**Comment by rigazilla** [comment]
@dmvolod it would be great a check that the new pods actually run the desired image, to detect issues like #600

File: test/e2e/main_test.go [hunk 1]
```diff
@@ -113,7 +113,7 @@ func TestOperatorUpgrade(t *testing.T) {
 	spec := DefaultSpec.DeepCopy()
 	name := strcase.ToKebab(t.Name())
 	spec.ObjectMeta.Name = name
-
+	spec.ObjectMeta.Namespace = tconst.Namespace
```
Comment (dmvolod) [inline test/e2e/main_test.go:116]: ObjectMeta is redundant
```suggestion
	spec.Namespace = tconst.Namespace
```

Comment (dmvolod) [reply test/e2e/main_test.go:116]: Or we can define namespace globally for all specs in ObjectMeta struct

**Comment by dmvolod** [comment]
> @dmvolod it would be great a check that the new pods actually run the desired image, to detect issues like #600

Yes, sure. Will add this logic

**Comment by dmvolod** [comment]
Fixed, @rigazilla please rivew

**Comment by rigazilla** [comment]
it seems that DEFAULT_IMAGE is empty running the "TO" phase, cluster starts with "latest"

**Comment by dmvolod** [comment]
> it seems that DEFAULT_IMAGE is empty running the "TO" phase, cluster starts with "latest"

Thanks, fixed

**Comment by rigazilla** [comment]
Looks good now.
Great job @dmvolod !

**Comment by dmvolod** [comment]
Thanks for review, @rigazilla and @ryanemerson 
I'm merging PR, and we can always improve this scripts as they are not related to the production code

**Pull request merged**